Apps bindings support

larkox opened the pull request:
#### Summary
Add support for App bindings. Bindings are the way Apps relate with the user agent (in this case the mobile client) and allow adding elements to the UI.

In this PR we are mainly adding:
- New types and data structures to work with bindings and locations
- Mechanism to perform calls (actions generated from the client, like clicking the post menu item or running an app slash command).
  - Call response handling will be completed on future PRs.
- Mechanism to retrieve all the relevant bindings from the server
- Add locations (places in the UI that use a binding to show an interactive element) to the Post Menu, and Channel Menu.
- Simple synchronization method for bindings (fetch all bindings on each channel navigation)

Known caveats:
- Threads and Permalink views may show the bindings for a channel different to the one they are. To be solved in future PRs.
- Error handling yet to be done. Planned for future PRs.

#### Ticket Link
First step on https://mattermost.atlassian.net/browse/MM-30613

mickmister reviewed (approved):
Awesome work @larkox! No issues found 👍

I've noted some TODOs that should be follow up tasks for the team:

- Error handling. Showing error strings from call responses to the user [Ticket](https://mattermost.atlassian.net/browse/MM-31666)
- Implement apps modals on mobile [Ticket](https://mattermost.atlassian.net/browse/MM-31520)
- Improve syncing of bindings. Currently every channel switch. [Ticket](https://mattermost.atlassian.net/browse/MM-31665)
    - Related ticket - Frontend should store apps bindings for multiple channels at once [Ticket](https://mattermost.atlassian.net/browse/MM-31664)

enahum commented:
@larkox can you please add in the description what App bindings are, or perhaps add more context? the ticket in itself does not say anything about it either

larkox commented:
@enahum Updated the description of the ticket with all the information. Feel free to ask if you have any other question.

larkox commented:
Updated the PR with two details:

- Added the "sendEphemeral" function, to return messages and errors from the app.
- Only show the apps post options on the main channel, not on threads, searches or permalink views.

levb commented on app/mm-redux/client/client4.ts:
```diff
@@ -3043,6 +3044,26 @@ export default class Client4 {
+    }
+
+    getAppsBindings = async (userID: string, channelID: string) => {
+        return this.doFetch(
```
Can we file a ticket about the (http) cache-ability of this request?

larkox replied:
A different one to this one? https://mattermost.atlassian.net/browse/MM-31123

larkox commented on app/mm-redux/selectors/entities/apps.ts:
```diff
@@ -0,0 +1,16 @@
+
+    if (location) {
+        const headerBindings = state.entities.apps.bindings.filter((b) => b.location === location);
+        return headerBindings.reduce((accum: AppBinding[], current: AppBinding) => accum.concat(current.bindings || []), []);
```
Not really. At this moment (no cache on client side) these will be fetch just once on each channel navigation, and they are potentially different each time.

What they could use is better naming, since these are not headerBindings. Just bindings.

larkox commented on app/mm-redux/types/apps.ts:
```diff
@@ -0,0 +1,182 @@
+    complete: string;
+}
+
+export type AutocompleteElement = AppField;
```
It will be used for slash commands. It is mainly to be more intention revealing in the following type definitions.
